### ISSUE TITLE
Guards narrow to `never` when starting from `unknown`

### ISSUE THREAD

**Issue by TkDodo** [remeda/remeda#144]
all guards narrow to `never` when you have an input that is of type `unknown`.

Given:

`x: unknown`

calling:

`isBoolean(x)`

will result in `x` having the the type `never`.

I would expect `x` to be of type `boolean` then. Note that it works fine if you start from e.g. `boolean | string`

@digimuza maybe you can help us out here?

**Comment by TkDodo** [comment]
I can see that you tried to make that better in [ts-prime](https://github.com/digimuza/ts-prime/blob/master/src/guards.ts#L57-L61) via a `type DefinitelyBoolean`, and the narrowing seems to work there, but it requires a `ts-ignore` 🤔

**Comment by TkDodo** [comment]
I don't know why we can't just do:

```
export function isBoolean(data: unknown): data is boolean {
```

am I missing something here? maybe @lstkz knows 😅  Maybe this doesn't work for the other guards like isObject and isArray ...

**Comment by lstkz** [comment]
@TkDodo 
It seems to work for me

![image](https://user-images.githubusercontent.com/4478684/129574286-f6c70d66-710e-464f-859c-a358bc91d7cd.png)

```ts
export function isBoolean(data: unknown): data is boolean {
  return true;
}

const x: unknown = null!; 
if (isBoolean(x)) {
  const z = x;
} 
```

Ts version: 4.4.0-dev

**Comment by TkDodo** [comment]
@lstkz yes, that works, but this is not how our current implementation looks. it's:

```
export function isBoolean<T>(data: T): data is Extract<T, boolean> {
```

and extracting `boolean` from `unknown` yields `never`.

I think for the boolean guard, we can use the `data is boolean` approach, but I'm not sure this will work for other guards, like array. If you pass in a specific array, you'd likely want it to narrow to that specific array and not `Array<any>`, so we can't easily get rid of the generic.

**Comment by TkDodo** [comment]
so ideally:

- `isArray(x)` where `x: unknown` should yield `Array<unknown>`

ts-prime does this with `DefinitelyArray<T>`:

```
type DefinitelyArray<T extends unknown> = Extract<
  T,
  Array<any> | ReadonlyArray<any>
> extends never
  ? ReadonlyArray<unknown>
  : Extract<T, Array<any> | ReadonlyArray<any>>;
```

which works for `unknown` but a) needs a ts-ignore and b) doesn't work for other cases. For example, if you pass in a `Promise<any>` to `isArray()`, we currently get `never`, which is most correct, but that approach yields `Promise<any> & ReadonlyArray<unknown>` which is super weird.

ideally, we'd keep everything as-is, but return `ReadonlyArray<unkonwn>` if the input is `unknown`

**Comment by TkDodo** [comment]
Thinking about this some more, I'm inclined to go with the solution that `ts-prime` has, even if i means adding `ts-ignores`.

There, `isString` correctly narrows  `unknown` to `string`, while also narrowing e.g. `"foo" | null` to `"foo"`. This looks awesome from a consumer perspective. What do you think?
